### ISSUE TITLE
API questions: decode and encode?

### ISSUE THREAD

**Issue by hannesm** [dinosaure/unstrctrd#4]
From your article (https://blog.x25519.net/2020-03-31--release-cycle-email.html), I understand this library main motivation is to handle the folding whitespace. Is this correct? With @linse I developed an iCalendar parser, where we encountered as well a folding whitespace, and decided to use a regular expression (https://github.com/roburio/icalendar/blob/v0.1.3/src/icalendar.ml#L1121-L1123) for handling it.

Now, the question arises: should we use unstrctrd instead? it would be convenient for us if there would be a `decode : string -> string` function -- i.e. something that (a) cannot fail (where could it?) and (b) strips the folding whitespaces away (by removing the newline and beginning space(s)). I was not able to find such a function.

Then, for the output side "max 75 characters" is required, I was unable to find such a function (again, maybe `encode : string -> max_width:int -> string`) -- would it be possible to integrate such a function (or is it present, and just me unable to find it)?

If you're curious about iCalendar and their Content-line, have a look at [their specification](https://tools.ietf.org/html/rfc5545#section-3.1).

**Comment by dinosaure** [comment]
> it would be convenient for us if there would be a decode : string -> string function -- i.e. something that (a) cannot fail (where could it?) and (b) strips the folding whitespaces away (by removing the newline and beginning space(s)). I was not able to find such a function.

So, the "parser" can fail where the input does not properly finish with a CRLF like:
- `"Hello"`
- `"Hello\r"`
- `"Hello\r\n "`

Note that the current implementation does catch the `ocamllex` exception... I should make an issue about that. In other side, it should be possible to provide a function and automatically put a CRLF at the end of the input.

Then, a function `fold_fws` exists to delete all FWS token, so `to_utf_8_string <.> fold_fws` should be enough to return a free-FWS output.

> Then, for the output side "max 75 characters" is required, I was unable to find such a function (again, maybe encode : string -> max_width:int -> string) -- would it be possible to integrate such a function (or is it present, and just me unable to find it)?

This feature exists but only on `mrmime` with an internal (but pretty general and independent) encoder. If you are really interesting to use it, I can figure out how to integrate it into `unstrctrd`.

I think we can figure out about a next release :+1: with what you want!

**Comment by dinosaure** [comment]
> Now, the question arises: should we use unstrctrd instead?

Then, `unstrctrd` wants to provide a stream API mostly (and be able to provide an `angstrom` parser). It depends on what you want but your regular expression is enough if you don't want to provide a stream API (which is required by `mrmime` in my case).

**Comment by hannesm** [comment]
thanks for your answer, @dinosaure -- indeed we use the regular expression conversion as a pre-processing step, and afterwards use angstrom for parsing -- so if there's angstrom integration (which means we'll only need to read each byte once), that'd even be more appealing. would you mind to point me at the Angstrom combinator to use?

**Comment by dinosaure** [comment]
Yes, `unstrctrd` package provides a sub-package `unstrctrd.parser` and the module `Unstrctrd_parser` which gives you the expected _combinator_. The parser expect a `bytes` as an argument and as an internal buffer used by `ocamllex` (something like `Bytes.create 0x7f` should be enough and fast - allocation into the minor-heap).
